From d6f2d099b29d03c4ebed7cc4827bdf7dda9b5108 Mon Sep 17 00:00:00 2001
From: 2023k8009929022-LiuYaoqing <liuyaoqing23@mails.ucas.ac.cn>
Date: Sun, 31 May 2026 16:23:05 +0800
Subject: [PATCH 02/13] B2: correct bit order of xip rdata

---
 src/device/SPI.scala | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/device/SPI.scala b/src/device/SPI.scala
index ca84ae85..878eabf1 100644
--- a/src/device/SPI.scala
+++ b/src/device/SPI.scala
@@ -153,9 +153,9 @@ class APBSPI(address: Seq[AddressSet])(implicit p: Parameters) extends LazyModul
       }
       is(xip_read) {
         when(apb_state === idle){
-          state := xip_hold
+          state := normal
           in.pready := true.B
-          in.prdata := apb_bus.prdata
+          in.prdata := Cat(apb_bus.prdata(7,0),apb_bus.prdata(15,8),apb_bus.prdata(23,16),apb_bus.prdata(31,24))
         }
       }
       is(xip_hold) {
-- 
2.34.1

