From a3535eed04bdf0bee31de185b3cfb069f05f5483 Mon Sep 17 00:00:00 2001
From: 2023k8009929022-LiuYaoqing <liuyaoqing23@mails.ucas.ac.cn>
Date: Sat, 13 Jun 2026 21:11:57 +0800
Subject: [PATCH 07/13] B2: fix GPIO seg7 code

---
 src/device/GPIO.scala | 35 +++++++++++++----------------------
 1 file changed, 13 insertions(+), 22 deletions(-)

diff --git a/src/device/GPIO.scala b/src/device/GPIO.scala
index a176602a..28243bf3 100644
--- a/src/device/GPIO.scala
+++ b/src/device/GPIO.scala
@@ -33,29 +33,20 @@ class gpioChisel extends Module {
       "b00000011".U, // 0
       "b10011111".U, // 1
       "b00100101".U, // 2
-      "b00001101".U, // 3//not fixed
-      "b10011001".U, // 4//not fixed
-      "b01001001".U, // 5//not fixed
-      "b01000001".U, // 6//not fixed
-      "b00011111".U, // 7//not fixed
-      "b00000001".U, // 8//not fixed
-      "b00001001".U, // 9//not fixed
-      "b00010001".U, // A//not fixed
-      "b11000001".U, // B//not fixed
-      "b01100011".U, // C//not fixed
-      "b10000101".U, // D//not fixed
-      "b01100001".U, // E//not fixed
-      "b01110001".U  // F//not fixed
+      "b00001101".U, // 3
+      "b10011001".U, // 4
+      "b01001001".U, // 5
+      "b01000001".U, // 6
+      "b00011111".U, // 7
+      "b00000001".U, // 8
+      "b00001001".U, // 9
+      "b00010001".U, // A
+      "b11000001".U, // B
+      "b01100011".U, // C
+      "b10000101".U, // D
+      "b01100001".U, // E
+      "b01110001".U  // F
 
-      //debug
-      // "b00000001".U,
-      // "b00000010".U,
-      // "b00000100".U,
-      // "b00001000".U,
-      // "b00010000".U,
-      // "b00100000".U,
-      // "b01000000".U,
-      // "b10000000".U,
     )
     segs(hex)
   }
-- 
2.34.1

